README.txt
matlab-ruby
    http://matlab-ruby.rubyforge.org/
    jonathan.younger@lipomics.com
    A big thank you to Lipomics Technologies, Inc. http://www.lipomics.com for sponsoring this project.
    
== DESCRIPTION:
  
A Ruby interface to the MATLAB interpreted language.

== FEATURES:
  
* Data type conversion between MATLAB and Ruby String, Boolean, Nil and Numeric values
* Matrix, CellMatrix and StructMatrix helper classes for working with MATLAB matrices

== USAGE:

  require 'matlab'
  
  engine = Matlab::Engine.new
  engine.put_variable "x", 123.456
  engine.put_variable "y", 789.101112
  engine.eval "z = x * y"
  engine.get_variable "z"
  
  matrix = Matlab::Matrix.new(20, 400)
  20.times { |m| 400.times { |n| matrix[m, n] = rand } }
  engine.put_variable "m", matrix
  
  engine.save "/tmp/20_x_400_matrix"
  engine.close
  
  # May also use block syntax for new
  Matlab::Engine.new do |engine|
    engine.put_variable "x", 123.456
    engine.get_variable "x"
  end

== REQUIREMENTS:

* MATLAB
* GCC or some other compiler to build the included extension
* SWIG (If you want to recompile the SWIG wrapper)
* Mocha (For testing only)

== INSTALL:

Simply do the following, after installing MATLAB:

  * ruby setup.rb config
  * ruby setup.rb setup
  * ruby setup.rb install

Alternatively, you can download and install the RubyGem package for
matlab-ruby (you must have RubyGems and MATLAB installed, first):

  * gem install matlab-ruby

If you have MATLAB installed in a non-standard location, you can specify the location of the include and lib files by doing:

  * gem install matlab-ruby -- --with-matlab-include=/usr/local/matlab/extern/include \
     --with-matlab-lib=/usr/local/matlab/bin/glnx86

Also, the gem ships with the C source-code pre-built, so 
you do not need to have SWIG installed. However, if you have SWIG installed
and you want to generate the C file yourself, you can specify the
<code>--with-swig</code> option.

== LEGAL

MATLAB is a trademark of The MathWorks, Inc.

== LICENSE

matlab-ruby is licensed under the MIT License.

Copyright (c) 2007 Jonathan Younger <jonathan.younger@lipomics.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


History.txt
== 2.0.3 / 2008-05-19

* 2 bug fixes
  * Empty matrices are converted to empty ruby arrays.
  * Empty ruby arrays are converted to an empty MATLAB matrix

== 2.0.2 / 2008-05-19

* 1 bug fix
  * Empty matrices are properly converted to empty Matlab::Matrices instead of floats.

== 2.0.1 / 2008-02-21

* 1 minor enhancement
  * Added block usage to Matlab::Engine.new
  
== 2.0.0 / 2008-01-03

* 1 bug fix
  * Added NilClass to_ruby method.
* 3 changes
  * Removed dynamic variable name setters/getters in favor of put_variable, get_variable methods. 
  * 1 dimensional MATLAB matrices are now converted into a normal Ruby Arrays.
  * 1 dimensional MATLAB struct matrices are now converted into normal Ruby Hashes.
* 3 enhancements
  * MATLAB functions may now be executed directly from the engine instead of having to use eval_string.
  * Ruby Arrays are now converted into 1 dimensional MATLAB matrices.
  * Ruby Hashes are now converted into 1 dimensional MATLAB struct matrices.
  
== 1.0.1 / 2007-04-27

* 1 bug fix
  * Need to check for nil value during MATLAB to Ruby StructMatrix conversion otherwise it segfaults.

== 1.0.0 / 2007-04-06

* 1 major enhancement
  * Birthday!



test/test_engine.rb
require 'test/unit'
require 'matlab'
require 'matlab/engine'
require 'rubygems'
require 'mocha'

class Driver
end

class EngineTest_Init < Test::Unit::TestCase
  def test_new
    driver = mock()
    driver.expects(:open).with("matfoo")
    Driver.expects(:new).returns(driver)
    
    engine = Matlab::Engine.new("matfoo", :driver => Driver)
  end
  
  def test_new_with_invalid_driver
    exception = assert_raise(LoadError) do
      engine = Matlab::Engine.new("matfoo", :driver => "foo")
    end
    assert_equal "no such file to load -- matlab/driver/foo/driver", exception.message
    
    exception = assert_raise(LoadError) do
      engine = Matlab::Engine.new("matfoo", :driver => :bar)
    end
    assert_equal "no such file to load -- matlab/driver/bar/driver", exception.message
  end
  
  def test_new_with_native_driver
    begin
      require 'matlab_api'
      
      Matlab::Driver::Native::API.expects(:engOpen).with("matfoo")
      assert_nothing_raised do
        engine = Matlab::Engine.new("matfoo")
      end
    rescue LoadError => e
      exception = assert_raise(RuntimeError) do
        engine = Matlab::Engine.new("matfoo")
      end
      assert_equal "no driver for matlab found", exception.message
    end
  end
end

class EngineTest < Test::Unit::TestCase
  def setup
    @driver = mock()
    @handle = mock()
    @driver.expects(:open).with("matfoo").returns(@handle)
    Driver.expects(:new).returns(@driver)
    
    @engine = Matlab::Engine.new("matfoo", :driver => Driver)
  end
  
  def test_eval_string
    @driver.expects(:eval_string).with(@handle, "z = x * y")
    @engine.eval_string "z = x * y"
  end
  
  def test_close
    @driver.expects(:close).with(@handle)
    @engine.close
  end
  
  def test_get_variable
    @driver.expects(:get_variable).with(@handle, "x")
    @engine.get_variable("x")
  end
  
  def test_put_variable
    @driver.expects(:put_variable).with(@handle, "y", 123.456)
    @engine.put_variable("y", 123.456)
  end
  
  def test_method_missing
    @engine.expects(:put_variable).with("mr0_rand", 2)
    @engine.expects(:put_variable).with("mr1_rand", 4)
    @engine.expects(:eval_string).with("rand(mr0_rand, mr1_rand)")
    @engine.expects(:get_variable).with("ans").returns("The Answer")
    @engine.expects(:eval_string).with("clear mr0_rand mr1_rand")
    
    assert_equal "The Answer", @engine.rand(2, 4)
  end
end

test/driver/native/test_driver.rb
$:.unshift "../../../lib"
require 'test/unit'
require 'matlab'
require 'matlab/driver/native/driver'
require 'rubygems'
require 'mocha'

class DriverTest < Test::Unit::TestCase
  include Matlab::Driver::Native
  
  def setup
    @driver = Driver.new
  end
  
  def test_open
    API.expects(:engOpen).with("matfoo")
    @driver.open("matfoo")
  end
  
  def test_close
    API.expects(:engClose).with("engine")
    @driver.close("engine")
  end
  
  def test_eval_string
    API.expects(:engEvalString).with("engine", "string")
    @driver.eval_string("engine", "string")
  end
  
  def test_get_variable
    variable = mock(:to_ruby => "bar")
    API.expects(:engGetVariable).with("engine", "foo").returns(variable)
    assert_equal "bar", @driver.get_variable("engine", "foo")
  end
  
  def test_put_variable
    variable = mock(:to_matlab => "bar")
    API.expects(:engPutVariable).with("engine", "foo", "bar")
    @driver.put_variable("engine", "foo", variable)
  end
end

test/driver/native/test_conversions.rb
$:.unshift "../../../lib"
require 'test/unit'
require 'matlab'
require 'matlab/driver/native/conversions'
require 'rubygems'
require 'mocha'

class ConversionsTest < Test::Unit::TestCase
  def test_string_to_matlab_to_ruby
    assert_equal "foo", "foo".to_matlab.to_ruby
  end
  
  def test_true_to_matlab_to_ruby
    assert_equal true, true.to_matlab.to_ruby
  end
  
  def test_false_to_matlab_to_ruby
    assert_equal false, false.to_matlab.to_ruby
  end
  
  def test_nil_to_matlab_to_ruby
    assert_equal nil, nil.to_matlab.to_ruby
  end
  
  def test_float_to_ruby
    assert_nil nil.to_matlab.to_ruby
    assert_equal 12.345, 12.345.to_ruby
  end
  
  def test_numeric_to_matlab_to_ruby
    assert_equal 12345.0, 12345.to_matlab.to_ruby
    assert_equal 12.345, 12.345.to_matlab.to_ruby
  end
  
  def test_array_to_cell_matrix
    cell_matrix = Matlab::CellMatrix.new(7, 1)
    cell_matrix[0, 0] = true
    cell_matrix[1, 0] = "1"
    cell_matrix[2, 0] = "2"
    cell_matrix[3, 0] = ["3", nil, 5]
    cell_matrix[4, 0] = 6
    cell_matrix[5, 0] = 7
    cell_matrix[6, 0] = false
    
    assert_equal cell_matrix, [true, "1", "2", ["3", nil, 5], 6, 7, false].to_cell_matrix
  end
  
  def test_array_to_matlab_to_ruby
    assert_equal [1.0, nil, 3.0], [1, nil, 3.0].to_matlab.to_ruby
    assert_equal [1.0, "2", false, ["foo", "bar", "baz"]], [1, "2", false, ["foo", "bar", "baz"]].to_matlab.to_ruby
  end
  
  def test_hash_to_matlab_to_ruby
    assert_equal({"foo" => "bar"}, {"foo" => "bar"}.to_matlab.to_ruby) 
    assert_equal({"foo" => [1.0,2.0,3.0]}, {"foo" => [1,2,3]}.to_matlab.to_ruby) 
    assert_equal({"foo" => { "bar" => [1.0,2.0,3.0, [4.0,5.0,6.0]] }}, {"foo" => { "bar" => [1,2,3, [4,5,6]] }}.to_matlab.to_ruby)
  end
  
  def test_matlab_matrix_to_matlab_to_ruby
    matrix = Matlab::Matrix.new(3, 3)
    3.times { |m| 3.times { |n| matrix[m, n] = rand } }
    matrix[1, 1] = nil
    
    assert_equal matrix, matrix.to_matlab.to_ruby
  end
  
  def test_should_convert_a_0x0_matlab_matrix
    matrix = Matlab::Matrix.new(0, 0)
    result = matrix.to_matlab.to_ruby
    assert_equal [], result 
  end
  
  def test_should_convert_an_empty_ruby_array
    array = []
    assert_equal array, array.to_matlab.to_ruby
  end
  
  def test_matlab_cell_matrix_to_matlab_to_ruby
    cell_matrix = Matlab::CellMatrix.new(3, 3)
    3.times { |m| 3.times { |n| cell_matrix[m, n] = (n == 1 ? rand.to_s : rand) } }
    cell_matrix[0, 0] = true
    cell_matrix[1, 1] = nil
    cell_matrix[2, 2] = false
    
    assert_equal cell_matrix, cell_matrix.to_matlab.to_ruby
  end
  
  def test_matlab_struct_matrix_to_matlab_to_ruby
    cats = Matlab::StructMatrix.new(17, 1, "name")
    17.times { |m| 1.times { |n| cats[m, n]["name"] = "Kitty #{m}.#{n}" } }
    
    struct_matrix = Matlab::StructMatrix.new(17, 1, "name", "age", "married", "cats", "car")
    17.times do |m|
      1.times do |n|
        struct_matrix[m, n]["name"] = "Bob #{m}.#{n}"
        struct_matrix[m, n]["age"] = (rand * 100).to_i
        struct_matrix[m, n]["married"] = (rand > 0.5)
        struct_matrix[m, n]["cats"] = cats
        struct_matrix[m, n]["car"] = nil
      end
    end
    assert_equal struct_matrix, struct_matrix.to_matlab.to_ruby
  end
end

lib/matlab.rb
# Copyright (c) 2007 Jonathan Younger <jonathan.younger@lipomics.com>
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

require 'matlab/matrix'
require 'matlab/engine'

lib/matlab/engine.rb
module Matlab
  
  # The Engine class encapsulates a single connection to a MATLAB instance.
  # Usage:
  #
  #   require 'matlab'
  #
  #   engine = Matlab::Engine.new
  #
  #   engine.put_variable "x", 123.456
  #   engine.put_variable "y", 789.101112
  #   engine.eval "z = x * y"
  #   p engine.get_variable "z"
  #
  #   engine.close
  # 
  #   # May also use block syntax for new
  #   Matlab::Engine.new do |engine|
  #     engine.put_variable "x", 123.456
  #     engine.get_variable "x"
  #   end
  #
  # Values are sent to and from MATLAB by calling a method on the
  # engine with the variable name of interest.
  class Engine
    # The low-level opaque engine handle that this object wraps.
    attr_reader :handle

    # A reference to the underlying MATLAB driver used by this engine.
    attr_reader :driver
    
    # Create a new Engine object that connects to MATLAB via the given command
    def initialize(command = "matlab -nodesktop -nosplash", options = {})
      load_driver(options[:driver])
      
      @handle = @driver.open(command)
      
      if block_given?
        begin
          yield self
        ensure
          close
        end
      end
    end
    
    # Sends the given string to MATLAB to be evaluated
    def eval_string(string)
      @driver.eval_string(@handle, string)
    end
    
    # Put a value to MATLAB via a given name
    def put_variable(name, value)
      @driver.put_variable(@handle, name, value)
    end
    
    # Get a value from MATLAB via a given name
    def get_variable(name)
      @driver.get_variable(@handle, name)
    end
    
    # Call a MATLAB function passing in the arguments
    def method_missing(method_id, *args)
      method_name = method_id.id2name
      
      variable_names = []
      args.each_with_index do |arg, index|
        variable_names << variable_name = "mr#{index}_#{method_name}"
        put_variable(variable_name, arg)
      end
      
      eval_string("#{method_name}(#{variable_names.join(", ")})")
      result = get_variable("ans")
      eval_string("clear #{variable_names.join(" ")}")
      result
    end
      
    # Closes this engine
    def close
      @driver.close(@handle)
    end
    
    private
      # Loads the corresponding driver, or if it is nil, attempts to locate a
      # suitable driver.
      def load_driver(driver)
        case driver
          when Class
            # do nothing--use what was given
          when Symbol, String
            require "matlab/driver/#{driver.to_s.downcase}/driver"
            driver = Matlab::Driver.const_get(driver)::Driver
          else
            [ "Native" ].each do |d|
              begin
                require "matlab/driver/#{d.downcase}/driver"
                driver = Matlab::Driver.const_get(d)::Driver
                break
              rescue SyntaxError
                raise
              rescue ScriptError, Exception, NameError
              end
            end
            raise "no driver for matlab found" unless driver
        end

        @driver = driver.new
      end
  end
end


lib/matlab/driver/native/conversions.rb
require 'matlab_api'
require 'matlab/matrix'

class String
  # Converts the string to a MATLAB mxArray
  def to_matlab
    Matlab::Driver::Native::API.mxCreateString(self)
  end
end

class TrueClass
  # Converts true to a MATLAB logical scalar
  def to_matlab
    Matlab::Driver::Native::API.mxCreateLogicalScalar(self)
  end
end

class FalseClass
  # Converts false to a MATLAB logical scalar
  def to_matlab
    Matlab::Driver::Native::API.mxCreateLogicalScalar(self)
  end
end

class NilClass
  # Converts nil to MATLAB NaN
  def to_matlab
    Matlab::Driver::Native::API.mxGetNaN
  end
  
  def to_ruby
    nil
  end
end

class Float
  # Converts the value to nil if it is a MATLAB NaN
  def to_ruby
    to_s == nil.to_matlab.to_s ? nil : self
  end
end

class Numeric
  # Converts the value to a MATLAB double
  def to_matlab
    Matlab::Driver::Native::API.mxCreateDoubleScalar(self)
  end
end

class Array
  # Converts the array into a 1 Dimensional MATLAB numeric or cell matrix
  def to_matlab
    if empty?
      matrix = Matlab::Matrix.new(0, 0)
      matrix.to_matlab
    elsif all? { |value| value.kind_of?(Numeric) || value.nil? }
      matrix = Matlab::Matrix.new(size, 1)
    
      each_with_index do |value, index|
        matrix[index, 0] = value
      end
      matrix.to_matlab
    else
      to_cell_matrix.to_matlab
    end
  end
  
  # Flattens and converts the array to a 1 Dimensional Matlab::CellMatrix
  def to_cell_matrix
    cell_matrix = Matlab::CellMatrix.new(size, 1)
    
    each_with_index do |value, index|
      cell_matrix[index, 0] = value
    end
    cell_matrix
  end
end

class Hash
  # Converts the hash into a 1 Dimensional MATLAB struct matrix
  def to_matlab
    struct_matrix = Matlab::StructMatrix.new(1, 1, *keys)
    
    each do |(key, value)|
      struct_matrix[0,0][key] = value
    end
    
    struct_matrix.to_matlab
  end
end

module Matlab
  class Matrix
    # Converts the matrix into a MATLAB numeric matrix
    def to_matlab
      matrix = Matlab::Driver::Native::API.mxCreateDoubleMatrix(m, n, Matlab::Driver::Native::API::MxREAL)
      double_array = Matlab::Driver::Native::API::DoubleArray.new(m * n)
      
      index = 0
      n.times do |column_index|
        m.times do |row_index|
          double_array[index] = (@cells[row_index][column_index] ? @cells[row_index][column_index].to_f : nil.to_matlab)
          index += 1
        end
      end
      
      Matlab::Driver::Native::API.mxSetPr(matrix, double_array)
      matrix
    end
    
    # Creates a Matlab::Matrix or Ruby Array from a MATLAB numeric matrix
    def self.from_matlab(matrix)
      m = Matlab::Driver::Native::API.mxGetM(matrix)
      n = Matlab::Driver::Native::API.mxGetN(matrix)
      
      matlab_matrix = self.new(m, n)
      double_array = Matlab::Driver::Native::API::DoubleArray.frompointer(Matlab::Driver::Native::API.mxGetPr(matrix))
      
      index = 0
      n.times do |column_index|
        m.times do |row_index|
          matlab_matrix[row_index, column_index] = (Matlab::Driver::Native::API.mxIsNaN(double_array[index]) ? nil : double_array[index])
          index += 1
        end
      end
      
      if m == 1 || n == 1
        matlab_matrix.cells.flatten
      else
        matlab_matrix
      end
    end
  end
  
  class CellMatrix
    # Converts the matrix into a MATLAB cell matrix
    def to_matlab
      matrix = Matlab::Driver::Native::API.mxCreateCellMatrix(m, n)
      
      index = 0
      n.times do |column_index|
        m.times do |row_index|
          value = (@cells[row_index][column_index].nil? ? Matlab::Driver::Native::API.mxCreateDoubleScalar(nil.to_matlab) : @cells[row_index][column_index].to_matlab)
          Matlab::Driver::Native::API.mxSetCell(matrix, index, value)
          index += 1
        end
      end

      matrix
    end
    
    # Creates a Matlab::CellMatrix or Ruby Array from a MATLAB cell matrix
    def self.from_matlab(matrix)
      m = Matlab::Driver::Native::API.mxGetM(matrix)
      n = Matlab::Driver::Native::API.mxGetN(matrix)
      
      cell_matrix = self.new(m, n)
      
      index = 0
      n.times do |column_index|
        m.times do |row_index|
          value = Matlab::Driver::Native::API.mxGetCell(matrix, index).to_ruby
          cell_matrix[row_index, column_index] = (value.nil? || value.to_s == nil.to_matlab.to_s ? nil : value)
          index += 1
        end
      end
      
      if m == 1 || n == 1
        cell_matrix.cells.collect { |cell| cell.first }
      else
        cell_matrix
      end
    end
  end
  
  class StructMatrix
    # Converts the matrix into a MATLAB struct matrix
    def to_matlab
      matrix = Matlab::Driver::Native::API.mxCreateStructMatrix(m, n, 0, nil)
      names.each { |name| Matlab::Driver::Native::API.mxAddField(matrix, name) }
      
      index = 0
      m.times do |row_index|
        n.times do |column_index|
          names.each do |name|
            value = (@cells[row_index][column_index][name].nil? ? Matlab::Driver::Native::API.mxCreateDoubleScalar(nil.to_matlab) : @cells[row_index][column_index][name].to_matlab)
            Matlab::Driver::Native::API.mxSetField(matrix, index, name, value)
          end
          index += 1
        end
      end

      matrix
    end
    
    # Creates a Matlab::StructMatrix or Ruby Hash from a MATLAB struct matrix
    def self.from_matlab(matrix)
      m = Matlab::Driver::Native::API.mxGetM(matrix)
      n = Matlab::Driver::Native::API.mxGetN(matrix)
      names = (0...Matlab::Driver::Native::API.mxGetNumberOfFields(matrix)).collect { |i| Matlab::Driver::Native::API.mxGetFieldNameByNumber(matrix, i) }
      
      struct_matrix = self.new(m, n, *names)
      
      index = 0
      m.times do |row_index|
        n.times do |column_index|
          names.each do |name|
            value = Matlab::Driver::Native::API.mxGetField(matrix, index, name)
            struct_matrix[row_index, column_index][name] = (value.nil? || Matlab::Driver::Native::API.mxIsEmpty(value) || value.to_ruby.to_s == nil.to_matlab.to_s ? nil : value.to_ruby)
          end
          index += 1
        end
      end
      
      if m == 1 && n == 1
        struct_matrix.cells.flatten.first
      else
        struct_matrix
      end
    end
  end
end

class SWIG::TYPE_p_mxArray_tag
  include Matlab::Driver::Native::API
  
  def to_ruby
    case
    when mxIsStruct(self)
      Matlab::StructMatrix.from_matlab(self)
    when mxIsCell(self)
      Matlab::CellMatrix.from_matlab(self)
    when mxIsChar(self)
      mxArrayToString(self)
    when mxIsLogical(self)
      mxIsLogicalScalarTrue(self)
    when (mxGetM(self) == 0 && mxGetN(self) == 0)
      []
    when (mxGetM(self) > 1 || mxGetN(self) > 1)
      Matlab::Matrix.from_matlab(self)
    when mxIsDouble(self)
      mxGetScalar(self)
    when mxIsEmpty(self)
      nil
    end
  end
end

lib/matlab/matrix.rb
module Matlab
  
  # The Matrix class is used for storing values that are sent to
  # or received from MATLAB numeric matrices
  # Usage:
  #
  #   require 'matlab'
  #
  #   matrix = Matlab::Matrix.new(20, 400)
  #   20.times { |m| 400.times { |n| matrix[m, n] = rand } }
  #
  class Matrix
    include Comparable
    
    attr_reader :m
    attr_reader :n
    attr_reader :cells
    
    # Creates a new Matrix with the given dimensions for row and column size
    def initialize(m, n)
      @m, @n = m, n
      @cells = Array.new(m) { Array.new(n) }
    end
    
    # Gets the value at the given row and column position
    def [](m, n)
      @cells[m][n]
    end
    
    # Sets the given value at the row and column position
    def []=(m, n, value)
      @cells[m][n] = value
    end
    
    # Enables comparisons of matrices to each other
    def ==(other)
      @cells == other.cells
    end
  end
  
  # The CellMatrix class is used for storing values that are sent to
  # or received from MATLAB cell matrices.
  # Usage:
  #
  #   require 'matlab'
  #
  #   cell_matrix = Matlab::CellMatrix.new(20, 400)
  #   20.times { |m| 400.times { |n| cell_matrix[m, n] = rand.to_s } }
  #
  class CellMatrix < Matrix; end
  
  # The StructMatrix class is used for storing hash values that are sent to
  # or received from MATLAB struct matrices.
  # Usage:
  #
  #   require 'matlab'
  #
  #   struct_matrix = Matlab::StructMatrix.new(17, 1, "name", "age", "married"
  #   17.times do |m|
  #     struct_matrix[m, 0]["name"] = "Bob #{m}"
  #     struct_matrix[m, 0]["age"] = (rand * 100).to_i
  #     struct_matrix[m, 0]["married"] = (rand > 0.5)
  #   end
  #
  #   p struct_matrix[16, 0]
  #
  class StructMatrix < Matrix
    attr_reader :names
    
    # Creates a new StructMatrix with the given dimensions for row and column size
    # and the names of the attributes
    def initialize(m, n, *names)
      super(m, n)
      @names = names
      
      # Populate the matrix with a hash using the names provided
      m.times do |row_index|
        n.times do |column_index|
          @cells[row_index][column_index] = names.inject({}) { |s,e| s.merge( { e.to_s => nil } ) }
        end
      end
    end
  end
end